src/operation/union/CascadedPolygonUnion.cpp

```diff
@@ -230,179 +231,35 @@ CascadedPolygonUnion::unionSafe(geom::Geometry* g0, geom::Geometry* g1)
-        disjointPolys.push_back(u.get());
-        ret = geom::util::GeometryCombiner::combine(disjointPolys);
-    }
-    else {
-        // TODO: could be further tweaked to only union with polysOn
-        //       and combine with polysOff, but then it'll need again to
-        //       recurse in the check for disjoint/intersecting
-        ret = geom::util::GeometryCombiner::combine(disjointPolys);
-        ret.reset(unionActual(ret.get(), u.get()));
-    }
-
-#if GEOS_DEBUG_CASCADED_UNION
-    check_valid(*ret, "unionUsingEnvelopeIntersection returned geom");
-#endif
-
-    return ret.release();
-}
-
-/* private */
-void
-CascadedPolygonUnion::extractByEnvelope(geom::Envelope const& env,
-                                        std::vector<geom::Geometry*>& sourceGeoms,
-                                        std::vector<geom::Geometry*>& intersectingGeoms,
-                                        std::vector<geom::Geometry*>& disjointGeoms)
-{
-    for(std::vector<geom::Geometry*>::iterator i = sourceGeoms.begin(),
-            e = sourceGeoms.end(); i != e; ++i) {
-        geom::Geometry* elem = *i;
-        if(elem->getEnvelopeInternal()->intersects(env)) {
-            intersectingGeoms.push_back(elem);
-        }
-        else {
-            disjointGeoms.push_back(elem);
-        }
-    }
-}
-
-/* private */
-void
-CascadedPolygonUnion::extractByEnvelope(geom::Envelope const& env,
-                                        geom::Geometry* geom,
-                                        std::vector<geom::Geometry*>& intersectingGeoms,
-                                        std::vector<geom::Geometry*>& disjointGeoms)
-{
-    for(std::size_t i = 0; i < geom->getNumGeometries(); i++) {
-        geom::Geometry* elem = const_cast<geom::Geometry*>(geom->getGeometryN(i));
-        if(elem->getEnvelopeInternal()->intersects(env)) {
-            intersectingGeoms.push_back(elem);
-        }
-        else {
-            disjointGeoms.push_back(elem);
-        }
-    }
-}
-
-/* private */
-geom::Geometry*
-CascadedPolygonUnion::extractByEnvelope(geom::Envelope const& env,
-                                        geom::Geometry* geom, std::vector<geom::Geometry*>& disjointGeoms)
-{
-    std::vector<geom::Geometry*> intersectingGeoms;
-    extractByEnvelope(env, geom, intersectingGeoms, disjointGeoms);
-
-    return geomFactory->buildGeometry(intersectingGeoms);
-}
 
 geom::Geometry*
 CascadedPolygonUnion::unionActual(geom::Geometry* g0, geom::Geometry* g1)
 {
-    return restrictToPolygons(std::unique_ptr<geom::Geometry>(g0->Union(g1))).release();
+    OverlapUnion unionOp(g0, g1);
+    geom::Geometry* unioned = unionOp.doUnion();
+    geom::Geometry* justPolys = restrictToPolygons(unioned);
+    delete unioned;
+    return justPolys;
 }
 
-std::unique_ptr<geom::Geometry>
-CascadedPolygonUnion::restrictToPolygons(std::unique_ptr<geom::Geometry> g)
+geom::Geometry*
+CascadedPolygonUnion::restrictToPolygons(geom::Geometry* g)
```

Comment:
Going from managed to raw pointers feels like a move in the wrong direction...

Reply:
Returned to previous state